test/watcherEx.test.ts

```diff
@@ -0,0 +1,104 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { newEnforcer } from 'casbin';
+import { RedisWatcherEx } from '../src/watcherEx';
+
+test('UpdateForAddpolicy', async () => {
+  const AddPolicyCallback = async (args: { sec: string; ptype: string; params: string[] }) => {
+    await e1.selfAddPolicy(args.sec, args.ptype, args.params);
+    expect(await e2.enforce('alice', '/dataset2/resource2', 'POST')).toBe(true);
+    expect(await e1.enforce('alice', '/dataset2/resource2', 'POST')).toBe(true);
+    await watcherEx.close();
+  };
+  const e1 = await newEnforcer('examples/authz_model.conf', 'examples/authz_policy.csv');
+  const e2 = await newEnforcer('examples/authz_model.conf', 'examples/authz_policy.csv');
+  const watcherEx = await RedisWatcherEx.newWatcherEx({ addPolicyCb: AddPolicyCallback }, 'redis://localhost:6379/5');
+  e1.setWatcherEx(watcherEx);
+  e2.setWatcherEx(watcherEx);
+  await e2.addPolicy('alice', '/dataset2/resource2', 'POST');
+});
+
+test('UpdateForRemovePolicy', async () => {
+  const RemovePolicyCallback = async (args: { sec: string; ptype: string; params: string[] }) => {
+    await e1.selfRemovePolicy(args.sec, args.ptype, args.params);
+    expect(await e2.enforce('alice', '/dataset1/resource1', 'POST')).toBe(false);
+    expect(await e1.enforce('alice', '/dataset1/resource1', 'POST')).toBe(false);
+    await watcherEx.close();
+  };
+  const e1 = await newEnforcer('examples/authz_model.conf', 'examples/authz_policy.csv');
+  const e2 = await newEnforcer('examples/authz_model.conf', 'examples/authz_policy.csv');
+  const watcherEx = await RedisWatcherEx.newWatcherEx({ removePolicyCb: RemovePolicyCallback }, 'redis://localhost:6379/5');
+  e1.setWatcherEx(watcherEx);
+  e2.setWatcherEx(watcherEx);
+  await e2.removePolicy('alice', '/dataset1/resource1', 'POST');
+});
+
+test('UpdateForRemoveFilteredPolicy', async () => {
+  const RemovePolicyFilteredCallback = async (args: { sec: string; ptype: string; field: number; fieldValues: string[] }) => {
+    await e1.selfRemoveFilteredPolicy(args.sec, args.ptype, args.field, args.fieldValues);
+    expect(await e2.enforce('alice', '/dataset1/resource1', 'POST')).toBe(false);
+    expect(await e2.enforce('alice', '/dataset1/any', 'POST')).toBe(false);
+    expect(await e1.enforce('alice', '/dataset1/resource1', 'POST')).toBe(false);
+    expect(await e1.enforce('alice', '/dataset1/any', 'POST')).toBe(false);
+    await watcherEx.close();
+  };
+  const e1 = await newEnforcer('examples/authz_model.conf', 'examples/authz_policy.csv');
+  const e2 = await newEnforcer('examples/authz_model.conf', 'examples/authz_policy.csv');
+  const watcherEx = await RedisWatcherEx.newWatcherEx({ removeFilteredPolicyCb: RemovePolicyFilteredCallback }, 'redis://localhost:6379/5');
+  e1.setWatcherEx(watcherEx);
+  e2.setWatcherEx(watcherEx);
+  await e2.removeFilteredPolicy(0, 'alice');
```

Comment:
I think this test is not working correctly. You should not `await` this line, but wait for `RemovePolicyFilteredCallback` instead after you have called `removeFilteredPolicy`

Reply:
`RemovePolicyFilteredCallback` will be called inside `removeFilteredPolicy` and it's an asynchronous function so we await.

Reply:
I don't think that's true. Callback function is not awaited in `node-casbin` either. It's called yes, but since it does not await for it, it won't ensure it's completion before function returns.

It might work now, but it can also break. Usually this message in tests is indicative that some async functions were still executing after the tests were completed.
https://github.com/node-casbin/redis-watcher/actions/runs/3083436438/jobs/4984362403#step:9:15

Reply:
@Sefriol should we await watcher callback in `node-casbin`?

Reply:
Nope. `node-casbin` should work normally even if the watcher fails. Waiting for the watcher could also be a huge performance bottleneck which we definitely do not want.